coalib/misc/Caching.py

```diff
@@ -0,0 +1,193 @@
+                        False otherwise.
+    """
+    log_printer.warn("The caching database is corrupted and will "
+                     "be removed. Each project will be re-cached "
+                     "automatically the next time coala is run.")
+    error_files = []
+    for file_name in files:
+        file_path = get_cache_data_path(log_printer, file_name)
+        cache_dir = os.path.dirname(file_path)
+        try:
+            os.remove(file_path)
+        except OSError:
+            error_files.append(file_name)
+
+    if len(error_files) > 0:
+        error_files = ", ".join(error_files)
+        log_printer.warn("There was a problem deleting the following "
+                         "files: " + error_files + ". Please delete "
+                         "them manually from " + cache_dir)
+        return False
+
+    return True
+
+
+def pickle_load(log_printer, filename, fallback=None):
+    """
+    Get the data stored in ``filename`` present in the coala user
+    config directory. Example usage:
+
+    >>> from pyprint.NullPrinter import NullPrinter
+    >>> from coalib.output.printers.LogPrinter import LogPrinter
+    >>> log_printer = LogPrinter(NullPrinter())
+    >>> test_data = {"answer": 42}
+    >>> pickle_dump(log_printer, "test_file", test_data)
+    >>> pickle_load(log_printer, "test_file")
+    {'answer': 42}
+    >>> pickle_load(log_printer, "nonexistant_file")
+    >>> pickle_load(log_printer, "nonexistant_file", fallback=42)
+    42
+
+
+    :param log_printer: A LogPrinter object to use for logging.
+    :param filename:    The name of the file present in the coala user config
+                        directory.
+    :param fallback:    Return value to fallback to in case the file doesn't
+                        exist.
+    :return:            Data that is present in the file, if the file exists.
+                        Otherwise the ``default`` value is returned.
+    """
+    filename = get_cache_data_path(log_printer, filename)
+    if not os.path.isfile(filename):
+        return fallback
+    with open(filename, "rb") as f:
+        try:
+            return pickle.load(f)
+        except (pickle.UnpicklingError, EOFError) as e:
+            delete_cache_files(log_printer, files=[filename])
+            return fallback
+
+
+def pickle_dump(log_printer, filename, data):
+    """
+    Write ``data`` into the file ``filename`` present in the coala user
+    config directory.
+
+    :param log_printer: A LogPrinter object to use for logging.
+    :param filename:    The name of the file present in the coala user config
+                        directory.
+    :param data:        Data to be serialized and written to the file using
+                        pickle.
+    """
+    filename = get_cache_data_path(log_printer, filename)
+    with open(filename, "wb") as f:
+        return pickle.dump(data, f)
+
+
+def add_to_changed_files(log_printer, project_dir, changed_files):
```

Comment:
How about giving the user an option to rebuild the cache? But that's basically purging the cache and restarting. Or should we verify if each file in the database exists at each run? The performance might take a hit.